ocaml/typing/typecore.ml

```diff
@@ -3428,58 +3458,93 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 mode_fun sargs ret
-             let open_args = (exp_mode, exp) :: open_args in
+             let open_args = (exp_mode, exp.exp_loc) :: open_args in
              let args = (lbl, Arg (exp, sort)) :: args in
-             (ty_ret, mode_ret, open_args, closed_args, args)
-         | Omitted { mode_fun; ty_arg; mode_arg; level; sort_arg } ->
+             (ty_ret, mode_ret, open_args, closed_args, underscore_args, args)
+         | Dummy { mode_fun; ty_arg; mode_arg; level; sort_arg } ->
+             let underscore_args = mode_arg :: underscore_args in
+             let args = (lbl, Dummy {mode_fun; ty_arg; mode_arg; level; sort_arg} ) :: args in
+             (ty_ret, mode_ret, open_args, closed_args, underscore_args, args)
+         | Omitted { mode_fun; ty_arg; mode_arg; level; sort_arg;} ->
              let arrow_desc = (lbl, mode_arg, mode_ret) in
              let ty_ret =
                newty2 ~level
                  (Tarrow (arrow_desc, ty_arg, ty_ret, commu_ok))
              in
              let new_closed_args =
                List.map
-                 (fun (marg, exp) ->
-                    submode ~loc:exp.exp_loc ~env ~reason:Other
+                 (fun (marg, loc) ->
+                    submode ~loc ~env ~reason:Other
                       marg (mode_partial_application expected_mode);
                     Value.regional_to_local_alloc marg)
                  open_args
              in
              let closed_args = new_closed_args @ closed_args in
              let open_args = [] in
-             let mode_closed_args = List.map Alloc.close_over closed_args in
-             let mode_partial_fun = Alloc.partial_apply mode_fun in
+             (* note that [mode_fun] is closed_over, not partial_apply *)
+             let all_closed = mode_fun :: closed_args @ underscore_args in
              let mode_closure, _ =
-               Alloc.newvar_above (Alloc.join
-                (mode_partial_fun:: mode_closed_args))
+               all_closed |> List.map Alloc.close_over |> Alloc.join |> Alloc.newvar_above
              in
              register_allocation_mode mode_closure;
              let arg = Omitted { mode_closure; mode_arg; mode_ret; sort_arg } in
              let args = (lbl, arg) :: args in
-             (ty_ret, mode_closure, open_args, closed_args, args))
-      (ty_ret, mode_ret, [], [], []) (List.rev args)
+             (ty_ret, mode_closure, open_args, closed_args, underscore_args, args))
+      (ty_ret, mode_ret, [], [], [], []) (List.rev args)
+  in
+  let closed_args =
+    if List.length underscore_args = 0 then closed_args
+    else
+    let new_closed_args =
+      List.map
+        (fun (marg, loc) ->
+          submode ~loc ~env ~reason:Other
+            marg (mode_partial_application expected_mode);
+          Value.regional_to_local_alloc marg)
+        open_args
+    in
+    funct_mode :: new_closed_args @ closed_args
+  in
+  let rec type_underscore_parameters closed_args = function
```

Comment:
What do you think about hoisting this out to a top-level function? That makes it clear which local variables it closes over. (In this case there are only two, the expected ones.) feel free to push back if you would rather not.

Reply:
I have no preference, so will do that.